File: src/shared/ipc-contracts/device-title.ts
/**
 * Wire contract for the `set_device_story_title` Tauri command. Mirror of
 * `src-tauri/src/ipc/dto/device_title.rs`.
 *
 * Input: the canonical `packUuid` the frontend holds from the inventory and
 * the raw title text. Rust normalizes (NFC + trim), validates (denylist +
 * ≤120) and persists it; provenance is Rust-owned (`user`), so it is NOT a
 * field of the input.
 *
 * Outcome: the stored title and its provenance (always `user`).
 */

import type { PackTitleSource } from "./device-library";

export interface SetDeviceStoryTitleInput {
  /** Canonical lowercase hyphenated pack UUID (8-4-4-4-12). */
  packUuid: string;
  /** Raw title text; Rust normalizes and validates it authoritatively. */
  title: string;
}

export interface DeviceStoryTitleDto {
  title: string;
  source: PackTitleSource;
}

const ALLOWED_INPUT_KEYS: ReadonlySet<string> = new Set(["packUuid", "title"]);
const ALLOWED_OUTCOME_KEYS: ReadonlySet<string> = new Set(["title", "source"]);
const TITLE_SOURCES: ReadonlySet<string> = new Set([
  "user",
  "official",
  "unofficial",
]);

/** Canonical lowercase hyphenated UUID (8-4-4-4-12). */
const CANONICAL_UUID_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

function hasOnlyAllowedKeys(
  value: Record<string, unknown>,
  allowed: ReadonlySet<string>,
): boolean {
  for (const k of Object.keys(value)) {
    if (!allowed.has(k)) return false;
  }
  return true;
}

/**
 * Client-side guard for the input. The `packUuid` is a Rust-issued
 * identifier, so a malformed value is a frontend bug — refused before the
 * round-trip. The `title` is only checked for being a non-blank string
 * here; Rust owns the authoritative normalization + validation (denylist,
 * length) and returns the canonical refusal on violation.
 */
export function isSetDeviceStoryTitleInput(
  value: unknown,
): value is SetDeviceStoryTitleInput {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (!hasOnlyAllowedKeys(c, ALLOWED_INPUT_KEYS)) return false;
  if (typeof c.packUuid !== "string" || !CANONICAL_UUID_PATTERN.test(c.packUuid)) {
    return false;
  }
  if (typeof c.title !== "string" || c.title.trim().length === 0) return false;
  return true;
}

/** Runtime guard for the `set_device_story_title` outcome. */
export function isDeviceStoryTitleDto(
  value: unknown,
): value is DeviceStoryTitleDto {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (!hasOnlyAllowedKeys(c, ALLOWED_OUTCOME_KEYS)) return false;
  if (typeof c.title !== "string" || c.title.length === 0) return false;
  if (typeof c.source !== "string" || !TITLE_SOURCES.has(c.source)) return false;
  return true;
}

File: src-tauri/src/ipc/dto/device_title.rs
use serde::{Deserialize, Serialize};

use crate::domain::device::title::PackTitle;

use super::device_library::PackTitleSourceDto;

/// Input accepted by the `set_device_story_title` Tauri command. The user
/// names a device story that no catalog recognizes (or renames one). The
/// frontend supplies the canonical `pack_uuid` it holds from the inventory
/// and the raw title text; Rust normalizes, validates and persists it.
/// `deny_unknown_fields` keeps the boundary authoritative.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SetDeviceStoryTitleInputDto {
    pub pack_uuid: String,
    pub title: String,
}

/// Outcome returned by a successful `set_device_story_title`: the stored
/// title and its provenance (always `user`). Mirror of
/// `src/shared/ipc-contracts/device-title.ts`.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeviceStoryTitleDto {
    pub title: String,
    pub source: PackTitleSourceDto,
}

impl DeviceStoryTitleDto {
    pub fn from_pack_title(title: PackTitle) -> Self {
        Self {
            title: title.title,
            source: title.source.into(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::device::title::PackTitleSource;

    #[test]
    fn input_accepts_canonical_camel_case_payload() {
        let dto: SetDeviceStoryTitleInputDto = serde_json::from_value(serde_json::json!({
            "packUuid": "abababab-abab-abab-abab-ababfac5562d",
            "title": "Mon histoire",
        }))
        .expect("deser");
        assert_eq!(dto.pack_uuid, "abababab-abab-abab-abab-ababfac5562d");
        assert_eq!(dto.title, "Mon histoire");
    }

    #[test]
    fn input_rejects_unknown_field() {
        let err = serde_json::from_value::<SetDeviceStoryTitleInputDto>(serde_json::json!({
            "packUuid": "x",
            "title": "y",
            "source": "official",
        }))
        .expect_err("must reject unknown field — provenance is Rust-owned");
        assert!(err.to_string().contains("source"));
    }

    #[test]
    fn input_rejects_snake_case_field() {
        serde_json::from_value::<SetDeviceStoryTitleInputDto>(serde_json::json!({
            "pack_uuid": "x",
            "title": "y",
        }))
        .expect_err("must reject snake_case");
    }

    #[test]
    fn outcome_serializes_with_camel_case_source_token() {
        let dto = DeviceStoryTitleDto::from_pack_title(PackTitle {
            title: "Mon histoire".into(),
            source: PackTitleSource::User,
            thumbnail: None,
        });
        let v = serde_json::to_value(&dto).expect("ser");
        assert_eq!(v["title"], "Mon histoire");
        assert_eq!(v["source"], "user");
    }
}
